pallets/delegation/src/migrations.rs

```diff
@@ -0,0 +1,741 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use codec::{Decode, Encode};
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData};
+
+use crate::*;
+
+/// A trait that allows version migrators to access the underlying pallet's
+/// context, e.g., its Config trait.
+///
+/// In this way, the migrator can access the pallet's storage and the pallet's
+/// types directly.
+pub trait VersionMigratorTrait<T> {
+	#[cfg(feature = "try-runtime")]
+	fn pre_migrate(&self) -> Result<(), &str>;
+	fn migrate(&self) -> Weight;
+	#[cfg(feature = "try-runtime")]
+	fn post_migrate(&self) -> Result<(), &str>;
+}
+
+/// Storage version of the delegation pallet.
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone, Encode, Eq, Decode, Ord, PartialEq, PartialOrd)]
+pub enum DelegationStorageVersion {
+	v1,
+	v2,
+}
+
+#[allow(dead_code)]
+impl DelegationStorageVersion {
+	/// The latest storage version.
+	fn latest() -> Self {
+		Self::v2
+	}
+}
+
+// All nodes will default to this, which is not bad, as in case the "real"
+// version is a later one (i.e. the node has been started with already the
+// latest version), the migration will simply do nothing as there's nothing in
+// the old storage entries to migrate from.
+//
+// It might get updated in the future when we know that no node is running this
+// old version anymore.
+impl Default for DelegationStorageVersion {
+	fn default() -> Self {
+		Self::v1
+	}
+}
+
+impl<T: Config> VersionMigratorTrait<T> for DelegationStorageVersion {
+	// It runs the right pre_migrate logic depending on the current storage version.
+	#[cfg(feature = "try-runtime")]
+	fn pre_migrate(&self) -> Result<(), &str> {
+		match *self {
+			Self::v1 => v1::pre_migrate::<T>(),
+			Self::v2 => Err("Already latest v2 version."),
+		}
+	}
+
+	// It runs the righ migration logic depending on the current storage version.
+	fn migrate(&self) -> Weight {
+		match *self {
+			Self::v1 => v1::migrate::<T>(),
+			Self::v2 => 0u64,
+		}
+	}
+
+	// It runs the right post_migrate logic depending on the current storage
+	// version.
+	#[cfg(feature = "try-runtime")]
+	fn post_migrate(&self) -> Result<(), &str> {
+		match *self {
+			Self::v1 => v1::post_migrate::<T>(),
+			Self::v2 => Err("Migration from v2 should have never happened in the first place."),
+		}
+	}
+}
+
+mod v1 {
```

Comment:
i would move this to it's own file. Especially with the tests in the end it becomes a bit overwhelming.

Reply:
Does this look good enough? https://github.com/KILTprotocol/mashnet-node/pull/226/commits/e96bc31fb6d8f4809f7a29616a363ce4b45af05f.

Reply:
Looks awesome! But maybe we keep the `migrations.rs` file instead of `migrations/mod.rs`? There is no clear rule for rust projects and both are equivalent (you can still have `migrations/v1.rs`). I personally favor `migrations.rs` since you won't have a thousand `mod.rs` files open where you don't know which one is which. 😅 
I researched this once and couldn't find a clear rule for that. So i guess do what you like more.